src/frontend/screens/DeviceNaming.tsx

```diff
@@ -0,0 +1,148 @@
+  header: {
+    id: 'screens.DeviceNaming.header',
+    defaultMessage: 'Add a name for your device',
+  },
+  description: {
+    id: 'screens.DeviceNaming.description',
+    defaultMessage:
+      'To create or join a new project you will have to name your device. In a future version of CoMapeo, you will be able to manage the projects you created by changing devices roles, deleting devices and much more.',
+  },
+  addName: {
+    id: 'screens.DeviceNaming.addName',
+    defaultMessage: 'Add Name',
+  },
+});
+
+export const DeviceNaming = ({
+  navigation,
+}: NativeStackScreenProps<DeviceNamingSceens, 'DeviceNaming'>) => {
+  const [name, setName] = React.useState('');
+  const [errorTimeout, setErrorTimeout] = useErrorTimeout();
+  const invalidName = name.length === 0 || name.length > 60;
+  const {formatMessage: t} = useIntl();
+
+  function setNameWithValidation(nameValue: string) {
+    if (nameValue.length > 60) {
+      setErrorTimeout();
+      return;
+    }
+    setName(nameValue);
+  }
+
+  function handleAddNamePress() {
+    if (invalidName) {
+      setErrorTimeout();
+      return;
+    }
+
+    navigation.navigate('Success', {deviceName: name});
+  }
+  return (
+    <KeyboardAvoidingView style={{width: '100%', height: '100%'}}>
+      <TouchableWithoutFeedback
+        style={styles.container}
+        onPress={Keyboard.dismiss}>
+        <View>
+          <CoMapeoText style={{alignSelf: 'center'}} />
+
+          <Text style={{marginTop: 20}}>{t(m.header)}</Text>
+          <TextInput
+            style={[
+              styles.textInput,
+              {borderColor: !errorTimeout ? LIGHT_GREY : RED},
+            ]}
+            value={name}
+            onChangeText={setNameWithValidation}
+            placeholderTextColor={LIGHT_GREY}
+            placeholder="Device Name"
+          />
+          <Text
+            style={{
+              alignSelf: 'flex-end',
+              color: errorTimeout ? RED : MEDIUM_GREY,
+            }}>
+            {`${name.length}/60`}
+          </Text>
+          <View style={styles.greyBox}>
+            <Text>{t(m.description)}</Text>
+          </View>
+        </View>
+
+        <Button fullWidth onPress={handleAddNamePress}>
+          {t(m.addName)}
+        </Button>
+      </TouchableWithoutFeedback>
+    </KeyboardAvoidingView>
+  );
+};
+
+function useErrorTimeout() {
+  const [errorTimeout, setErrorTimeout] = React.useState(false);
+  const timer = React.useRef<NodeJS.Timeout | undefined>();
+
+  React.useEffect(() => {
+    if (errorTimeout && !timer.current) {
+      timer.current = setTimeout(() => {
+        setErrorTimeout(false);
+        timer.current = undefined;
+      }, 1500);
+    }
+
+    return () => {
+      if (timer.current) {
+        clearTimeout(timer.current);
+        timer.current = undefined;
+      }
+    };
+  }, [errorTimeout]);
+
+  return [
+    errorTimeout,
+    () => setErrorTimeout(prevVal => (!prevVal ? true : prevVal)),
+  ] as const;
+}
```

Comment:
Think this could be implemented in a way that:

1. Avoids effects that rely on state changes and instead colocates them with explicit event handlers/callbacks
2. Generalizes the value that's ephemeral
3. Allows configurable time length

Ended up with this, which is similar to `useState` in terms of usage with a couple of extra things:

1. An option to configure duration before resetting to the initial value
2. A third element in the return value which can be used to reset to the initial state before the duration is reached (e.g. clearing an error state immediately based on some subsequent user interaction)

```ts
// initialValue and resetAfter usually won't change too much, as they are meant to be like configuration options, similar to how the built-in React hooks work
function useEphemeralState<T>(initialValue: T, resetAfter: number) {
  const [value, setValue] = React.useState<T>(initialValue);

  const timer = React.useRef<NodeJS.Timeout | undefined>();

  // There's probably something I'm missing here in terms of resetting some state or cleaning up when initialValue or resetAfter change, but not sure how consequential that is
  React.useEffect(() => {
    return () => {
      if (timer.current) {
        clearTimeout(timer.current);
        timer.current = undefined;
      }
    };
  }, []);

  function set(value: Parameters<typeof setValue>[0]) {
    setValue(value);
    
    // co-locate the restart/creation of the timer with the event handler/callback, based on React idioms
    if (timer.current) {
      clearTimeout(timer.current);
    }

    timer.current = setTimeout(() => {
      setValue(initialValue);
    }, resetAfter);
    

  }

  function reset() {
    setValue(initialValue);

    if (timer.current) {
      clearTimeout(timer.current);
      timer.current = undefined;
    }
  }

  return [value, set, reset] as const;
}
```

In this component, usage would be like this (although):

```ts
// `error` is inferred to be boolean here, so the explicit generic isn't technically needed in this case
const [error, setError] = useEphemeralState<boolean>(false, 1500);

function handleNamePress() {
  if (invalidName) {
    setError(true)
  }
  
  ...
}
```

Not adamant about using this change exactly as is, but I think it would be worth considering to avoid dealing with more complicated useEffects

Reply:
I agree, the useEffect is not ideal. But I think instead of implementing custom validation hooks we should look into libraries (eg. react hook forms). I am going to leave it as is, but earmark it as something we should address